free5gc-operator/controllers/smfdeployment_controller.go

```diff
@@ -0,0 +1,648 @@
+}
+
+func constructSMFNadName(templateName string, suffix string) string {
+	return templateName + "-" + suffix
+	// return "n4"
+}
+
+// getSMFNad returns NAD label string composed based on the Nx interfaces configuration provided in SMFDeploymentSpec
+func getSMFNad(templateName string, spec *workloadv1alpha1.SMFDeploymentSpec) string {
+	var ret string
+
+	// n6CfgSlice := getIntConfigSlice(spec.Interfaces, "N6")
+	// n3CfgSlice := getIntConfigSlice(spec.Interfaces, "N3")
+	n4CfgSlice := getIntConfigSlice(spec.Interfaces, "n4")
+	// n9CfgSlice := getIntConfigSlice(spec.Interfaces, "N9")
+
+	ret = `[`
+	intfMap := map[string][]workloadv1alpha1.InterfaceConfig{
+		// "n3": n3CfgSlice,
+		"n4": n4CfgSlice,
+		// "n6": n6CfgSlice,
+		// "n9": n9CfgSlice,
+	}
+	// Need to sort inftMap by key otherwise unitTests might fail as order of intefaces in intfMap is not guaranteed
+	inftMapKeys := make([]string, 0, len(intfMap))
+	for interfaceName := range intfMap {
+		inftMapKeys = append(inftMapKeys, interfaceName)
+	}
+	sort.Strings(inftMapKeys)
+
+	noComma := true
+	for _, key := range inftMapKeys {
+		for _, intf := range intfMap[key] {
+			newNad := fmt.Sprintf(`
+        {"name": "%s",
+         "interface": "%s",
+         "ips": ["%s"],
+         "gateways": ["%s"]
+        }`, constructSMFNadName(templateName, key), intf.Name, intf.IPv4.Address, *intf.IPv4.Gateway)
+			if noComma {
+				ret = ret + newNad
+				noComma = false
+			} else {
+				ret = ret + "," + newNad
+			}
+		}
+	}
+	ret = ret + `
+    ]`
+	return ret
+
+	// return `[
+	//     {"name": "free5gc-smf-1-n4",
+	//      "interface": "n4",
+	//     },
+	// ]`
+
+}
+
+// checkNADExists gets deployment object and checks "k8s.v1.cni.cncf.io/networks" NADs.
+// returns True if all requred NADs are present
+// returns False if any NAD doesn't exists in deployment namespace
+func (r *SMFDeploymentReconciler) checkSMFNADexist(log logr.Logger, ctx context.Context, deployment *appsv1.Deployment) bool {
+	smfAnnotations := []SMFAnnotation{}
+	annotationsString, ok := deployment.Spec.Template.GetAnnotations()["k8s.v1.cni.cncf.io/networks"]
+	if !ok {
+		log.Info("Annotations k8s.v1.cni.cncf.io/networks not found", "SMFDeployment.namespace", deployment.Namespace)
+		return false
+	}
+	if err := json.Unmarshal([]byte(annotationsString), &smfAnnotations); err != nil {
+		log.Info("Failed to parse SMFDeployment annotations", "SMFDeployment.namespace", deployment.Namespace)
+		return false
+	}
+	for _, smfAnnotation := range smfAnnotations {
+		u := &unstructured.Unstructured{}
+		u.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   "k8s.cni.cncf.io",
+			Kind:    "NetworkAttachmentDefinition",
+			Version: "v1",
+		})
+		key := client.ObjectKey{Namespace: deployment.ObjectMeta.Namespace, Name: smfAnnotation.Name}
+		if err := r.Get(ctx, key, u); err != nil {
+			log.Info(fmt.Sprintf("Failed to get NAD %s", smfAnnotation.Name), "SMFDeployment.namespace", deployment.Namespace)
```

Comment:
So, yeah, we need to change the method to return (bool, error). In this case, if the error is "not found", I think that's not an error but part of what the method is checking, correct? In that case, we would `return false, nil` and be done. But if it's a different error, then it indicates some failure and we should `return false, err` and let the caller figure out what to do about the error (most likely, it will be to set a failed Ready condition in the object status and surface the error to the user).

Here's an example of using the `IgnoreNotFound` - though you should look up exactly what it returns as I am not sure offhand: https://github.com/GoogleContainerTools/kpt/blob/main/porch/controllers/packagevariants/pkg/controllers/packagevariant/packagevariant_controller.go#L147